mplayerd.1
.TH "mplayerd" "1" "0.0.9" "Sig Lange" ""
.SH "NAME"
.LP 
mplayerd \- command driven daemon to control mplayer via tcp/ip.
.SH "SYNTAX"
.LP 
mplayerd [\fIoptions\fP]
.br 
.SH "DESCRIPTION"
.LP 
mplayerd is a threaded daemon designed to control mplayer. It features a tab completion CLI, built in command help, unique command name recognition, logging, command history, multiple instance support, and more.
.SH "OPTIONS"
.LP 
.TP 
\-h \-\-help
Print this help message

.TP 
\-fg \-\-foreground
Don't fork

.TP 
\-p \-\-port port
Specify port, default 7400

.TP 
\-l \-\-logfile file
Log to file, default off

.TP 
\-v \-\-version
display version and exit

.TP 
\-d \-\-debug
debug, can specify multiple times

.TP 
\-c \-\-config
specify config file
.SH "FILES"
.LP 
\fImplayerd.conf\fP 

Two configuration sections [mplayer] and [daemon] apply to mplayer and the daemon itself, respectively.
.P
mplayerd configuration file directives:
.P
\fB[mplayer] section\fR
.br
.TP
path = mplayer
The path to mplayer, this needs to be the absolute path if mplayer is not in your $PATH.
Default: 'mplayer'

.TP
vo = driver
Set the video output device to driver. To get a list of valid video output devices on your setup run 'mplayer \-vo help' Most command are xv, sdl, and x11.
Default: 'x11'

.TP
max_instances = X
maximum number of instances that mplayerd can create.
Default: 4
.br
.br
.P
\fB[daemon] section\fR
.TP
port = 7400
Port to listen and accept connections.
Default: 7400

.TP
allow = A.B.C.D A.B.C.D
Specifies who is able to connect to the daemon. Only specified octets are evaluated against the connecting IP.
So, specifying 123. would allow 123.0.0.0/8 to connect. Seperate multiple IP addresses with a space.
Default: ''

.TP
root = /
Locks a client into a particular directory, they are unable to navigate outside of this directory.
Default: '/'

.TP
default_home = /
Starting path when a client initially connects. The path specified here is relative to root.
Default: '/'

.SH "EXAMPLE CONFIGURATION"
.LP
# sample configuration file for mplayerd
.br
# suppose you wanted to use xv as your video out device and allow only 2 instances of mplayer to be created, the configuration
.br
# could be something like the following:

.P
[mplayer]
.br
#default 'mplayer'
.br
path = mplayer
.br
.br
vo = xv
.br
flags =
.br
#default 4
.br
max_instances = 2
.P
[daemon]
.br
#default 7400
.br
port = 7400
.br
allow = 127.0.0.1
.br
.br
#default /
.br
root = /
.br
.br
# relative to root
.br
default_home = /
.br


.SH "MPLAYERD COMMANDS"
.LP 

.P
\fBSyntax:\fR cd $directory
.br
\fBDescription:\fR change directory
.br
Change to directory $directory

.P
\fBSyntax:\fR pwd
.br
\fBDescription:\fR print working directory
.br
Print current working directory

.P
\fBSyntax:\fR ls [ $directory ]
.br
\fBDescription:\fR list files
.br
list files in current directory or $directory

.P
\fBSyntax:\fR version
.br
\fBDescription:\fR display version
.br
displays the version of mplayerd

.P
\fBSyntax:\fR status [ $instance ]
.br
\fBDescription:\fR display status
.br
display current status of movie/mplayerd

.P
\fBSyntax:\fR quit
.br
\fBDescription:\fR close session
.br
close your session

.P
\fBSyntax:\fR shutdown $string
.br
\fBDescription:\fR shutdown mplayerd
.br
shut down mplayerd, $string must be 'now'

.P
\fBSyntax:\fR help $string
.br
\fBDescription:\fR display help
.br
type help $string for more information on command $string

.P
\fBSyntax:\fR load [ $instance ] $file
.br
\fBDescription:\fR load a file
.br
load $file into mplayer, use quotes for file names with spaces

.P
\fBSyntax:\fR pause [ $instance ]
.br
\fBDescription:\fR pause/unpause
.br
pause/unpause playback -- alias to play

.P
\fBSyntax:\fR fullscreen [ $instance ]
.br
\fBDescription:\fR toggle fullscreen
.br
toggle fullscreen for media playback

.P
\fBSyntax:\fR stop [ $instance ]
.br
\fBDescription:\fR stop (close) movie
.br
stop and close the playing movie

.P
\fBSyntax:\fR osd [ $instance ] $integer
.br
\fBDescription:\fR on screen display level
.br
change the OSD (On Screen Display) level $integer
.br
0 - no OSD
.br
1 - seek bar
.br
2 - seek bar + seconds
.br
3 - seek bar + seconds + time remaining

.P
\fBSyntax:\fR seek_percent $instance $integer
.br
\fBDescription:\fR seek by percent
.br
seek to percent $integer of the movie
.br
not all media will support seeking

.P
\fBSyntax:\fR seek_absolute $instance $integer
.br
\fBDescription:\fR seek by seconds
.br
seek to an absolute position $integer in seconds
.br
not all media will support seeking

.P
\fBSyntax:\fR seek_relative $instance $integer
.br
\fBDescription:\fR seek relatively by seconds
.br
seek from current position by $integer seconds
.br
not all media will support seeking

.P
\fBSyntax:\fR literal [ $instance ] $string
.br
\fBDescription:\fR send mplayer a command
.br
send mplayer a command directly
.br
enclose entire command $string in double quotes
.br
see man mplayer(1) for more information about SLAVE MODE PROTOCOL

.P
\fBSyntax:\fR volume $instance $integer
.br
\fBDescription:\fR adjust the volume down/up
.br
adjust the volume down/up
.br
$integer is 0 for down and 1 for up

.P
\fBSyntax:\fR mute [ $instance ]
.br
\fBDescription:\fR mute volume
.br
mute the volume

.P
\fBSyntax:\fR kill $integer
.br
\fBDescription:\fR kill session
.br
kill user session $integer

.P
\fBSyntax:\fR history
.br
\fBDescription:\fR display command history
.br
display previous command history

.P
\fBSyntax:\fR reset [ $instance ]
.br
\fBDescription:\fR reset mplayerd
.br
reset mplayer if something should go wrong

.P
\fBSyntax:\fR qload [ $instance ] $integer
.br
\fBDescription:\fR quick load
.br
load file $integer from current directory

.P
\fBSyntax:\fR clear
.br
\fBDescription:\fR clear terminal
.br
clear the terminal screen

.P
\fBSyntax:\fR instances
.br
\fBDescription:\fR display instances
.br
displays a list of available instances

.P
\fBSyntax:\fR new
.br
\fBDescription:\fR create new instance
.br
create a new instance, prints the new instance id

.P
\fBSyntax:\fR delete $instance
.br
\fBDescription:\fR delete instance
.br
deletes specified instance

.P
\fBSyntax:\fR who
.br
\fBDescription:\fR display who's online
.br
display who is currently connected to mplayerd
.br
displays ID, IP and CWD

.P
\fBSyntax:\fR arguments [ $instance ] $string
.br
\fBDescription:\fR change arguments for mplayer
.br
change/show arguments for instance, enclose arguments in double quotes


.SH "EXAMPLES"
.LP 
To run this program the standard way type:
.LP 
$ mplayerd
.LP 
To keep it in the foreground with a good amount of debugging, use:
.LP 
$ mplayerd \-fg \-d \-d \-d 


.SH "AUTHORS"
.LP 
Sig Lange <exonic@signuts.net>
.SH "SEE ALSO"
.LP 
mplayer(1)


CHANGELOG
- new id3 command to display id3 tag information on files (requires mp3info)
- fixed a rediculous bug for testing purposes that disabled mplayer_message() processing!

2005-02-23 mplayerd 0.1.20
--------------------------
- after many many revisions, this version should be loads more stable
- fixed some #ifdef's so you can have instance support without tab completion
- stopped setting pause state when mplayer prints '=== PAUSED ===' type output.
- got rid of a few more malloc's
- updated command history structure
- fixed some logic problems in the way mplayer_unload() is being called 
- forgot a mutex_lock calling get_instance in mplayer_message*

2004-06-01 mplayerd 0.1.17
--------------------------
- added a better check which looks for :// in file names to determine if it's a URL. This fixes mplayerd unable to load
  streams
- better status return codes for various commands output. ( instance, arguments, delete, add, dvd commands)
- fixed include/build error when DEBUG isn't defined
- cleaned CMD_LS and added a directory suffix '/' (to allow frontends to determine file types)
- added file stat() call to mplayer_load() to load only files

2004-04-29 mplayerd 0.1.16
--------------------------
- non blocking sockets dropped
- cleaned up / optimized tab completion code
- new dvd command to load titles from dvd

2003-12-01 mplayerd 0.1.14
--------------------------
- added two defines to enable/disable tab completion and instance support at compile time


2003-11-10 mplayerd 0.1.12
--------------------------
- max_instances config directive added, limiting the max number of instances
- setup defaults, so operation w/o config file is possible
- various tab completion bugs
- mplayerd man page
- arguments command to change mplayer flags on the fly
- tons of bugs fixes and code clean-ups

2003-10-05 mplayerd 0.0.9
-------------------------
- multiple instance support
- a locking contention issues fixed
- various bug/typo fixes
- fixed a .. directory traversal bug in tab completion and load command
- all commands should properly deal with the instance command
- under extreme load mplayer_command(..) timesout on write, increased timeout to 0.75
  timeout read/write for mplayer now in config.h
- fixed bugs while sending commands to a paused mplayer
- most commands now have "smarter" tab completion

2003-10-02 mplayerd 0.0.8
-------------------------
- load will now act as qload if the file isn't found
- fixed some bugs in the linked-list ll_add(..) !! Man!?
- tab completion code more organized


2003-09-30 mplayerd 0.0.7
-------------------------
- added xstrdup() command that calls xmalloc() instead
- added "qload" command
- '?' help special key improved
- added URL support -- Thanks to Michael Joosten <joost@xxxxxxxx>

2003-09-29 mplayerd 0.0.6
-------------------------
- moved various #defines to inc/config.h
- added -c --config flags to specify config file
- added flags directive to config file for additional mplayer flags
- added default_home directive to config file for default home for clients

2003-09-26 mplayerd 0.0.5
-------------------------
- up arrow / down arrow history implemented
- applied patch from Michael Joosten <joost@xxxxxxxx> which fixed the following:
	* fixed a bug where mplayer exit'ed and went wild on a bad descriptor in the mplayer message thread
	* status is updated to stop when mplayer exits normally now
- took out mplayerd.conf from install target of Makefile, do "make install_config" instead

2003-09-25 mplayerd 0.0.4
-------------------------
- incorporated help args into help, cleaner 'help [cmd]' output, ommitted short
  description in 'help [cmd]' output too
- fixed off by one error dealing with RECV_BUF

2003-09-18 mplayerd 0.0.3
-------------------------
- various functions that dealt with the client_list struct now simply pass the index
  of the client instead of socket, id, etc.
- mplayer_message thread launched at startup
- ADDED: config.h .. now contains config options
- various cosmetics:
	mplayer_help()	->	mplayerd_help()
	eliminated ip_access.[ch], moved functions to client.c and parse_config.c
	config.[ch] 	->	parse_config.[ch]
	got rid of client_sock_id(..) -- not used
- debug_msg(..) funciton implemented, DBG define not really used
- cleaner logging, more organized debug levels
- added xmemory.c, eventually will support mmap & malloc
- more versatile help command
- fixed a crucial bug in the fullscreen command attempting to obtain a lock twice. Ugh!

2003-09-17 mplayerd 0.0.2
-------------------------
- converted to threading, too much going on to switch back and forth
- added xmalloc(..) eventual options to use mmap or malloc
- clientRequest renamed to client_request
- clientSend renamed to client_send
- loadDirectory renamed to load_dir, and added a max argument to load_dir(..)
- better parsing in message thread, mpx now has audio_sec and video_sec
- "status" command output updated to reflect the latter
- SIGPIPE not ignored at times anymore & mplayer_dead() removed
- various bug fixes and clean ups

2003-??-?? mplayerd 0.0.1
-------------------------
 ( This is from memory )

- compiled one central changelog
- turned sockets all nonblocking
- clientSend(..) function uses select(..)
- clientRequets(..) checks return code of clientSend(..)
- Various bug fixes, code clean up
- #define OUTPUT_DONE "\xA0" added, for front ends
- command history implemented
- more organized command list structure
- too many changes to remember!


src/mplayerd.c
/*
 * mplayerd
 *
 * Website: http://mplayerd.sourceforge.net/
 * Secondary Website: http://signuts.net/projects/id/59
 *
 * http://www.signuts.net/
 * exonic@signuts.net
 * 
 *
 * Copyright (C) 2003, Sig Lange <exonic@signuts.net>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 *
 *
 */
#include "config.h"
#include "mplayerd.h"
#include "mplayer_command.h"
#include "client.h"
#include "client_request.h"
#include "parse_config.h"
#include "debug.h"
#include "xmemory.h"
#include "instance.h"

int debug = 0;
char *logfile;
int logging = 0, logging_fd;
int doFork = 1;
client_list clients[MAX_CLIENTS];

mpd_config *config = NULL;

char config_file[1024];

pthread_t main_pthread;
pthread_t mplayer_message_pthread;

/* pointer to current instance */
int instance_count = 0;


struct ll_list *mp_instances = NULL;

pthread_mutex_t instances_mutex = PTHREAD_MUTEX_INITIALIZER;

extern cmd_list command_list[];
int listen_sock;

int main(int argc, char *argv[]) {
	int new_client, i, clinum;
	struct sockaddr_in cl_addr;
	int cl_addrlen;
	int pid;
	char *p;

	init_clients();
	cl_addrlen = sizeof(cl_addr);

#ifdef DEBUG_MEMORY
	ll_mem_init();
#endif

	main_pthread = pthread_self();

	get_config_file_name(argc, argv, config_file);

	/* parse config file -- command line options override config directives */
	if ((config = parse_config(config_file)) == NULL) {
		printf("Couldn't parse config file '%s'! Exiting...\n", config_file);
		exit(1);
	}

	if (parse_argv(argc, argv) == -1) exit(1);

	if (config->default_home[0] == 0) {
		p = getcwd(NULL, 0);
		strcpy(config->default_home, p);
		xfree(p);
	}

	if (strcmp(config->default_home, "$HOME") == 0) {
		strcpy(config->default_home, getenv("HOME"));
	}

	if (logging) {
		logging_fd = open(logfile, O_WRONLY|O_APPEND|O_CREAT, S_IRWXU);
		if (logging_fd == -1) {
			printf("WARNING: Logging disabled!\n");
			logging = 0;
		}
	}

	if ( (listen_sock = server_setup()) == -1) exit(1);

	mp_instances = ll_new();
	if (!mp_instances) {
		printf("mplayer instance couldn't be init'd... failed! Exiting..\n");
		exit(1);
	}

	new_instance();

	if (doFork) {   
		printf("forking into background...\n");

		pid = fork();
		if (pid == -1) {
			perror("fork");
			exit(1);
		}

		if (pid) {
			exit(0);
		}
	}

	signal(SIGINT, (void *) mplayerd_quit);
	signal(SIGTERM, (void *) mplayerd_quit);
	signal(SIGPIPE, SIG_IGN);


	if (pthread_create(&mplayer_message_pthread, NULL, (void *) mplayer_message_thread, NULL)) return -1;

	debug_msg(0, "mplayerd %s started on port %d with pid %d\n", MPLAYERD_VERSION, config->port, getpid());
	debug_msg(1, "main thread id %ld\n", main_pthread);


	for( ;; ) {

		new_client = accept(listen_sock, (struct sockaddr *)&cl_addr, &cl_addrlen);

		debug_msg(0, "New connection from %s\n", (char *)inet_ntoa(cl_addr.sin_addr));

		if (new_client > 0) {

			debug_msg(0, "%s : connected.\n", (char *)inet_ntoa(cl_addr.sin_addr));

			/* make sure they're allowed to connect */
			if (check_ip((char *)inet_ntoa(cl_addr.sin_addr), config->allow_ips) == 0) {
				shutdown(new_client, 2);
				close(new_client);
				debug_msg(0, "%s : Access denied.\n", (char *)inet_ntoa(cl_addr.sin_addr));

			} else {

				clinum = -1;
				for(i=0; i<MAX_CLIENTS; i++) {
					if (clients[i].socket == -1) {
						clinum = i;
						break;
					}
				}

				if (clinum == -1) {
					debug_msg(0, "%s : No more connections allowed. disconnected.\n", (char *)inet_ntoa(cl_addr.sin_addr));

					send(new_client, "Sorry. No more connections allowed!\n", 37, MSG_NOSIGNAL);
					shutdown(new_client, 2);
					close(new_client);

				} else {

					clients[i].id = clinum;
					clients[i].socket = new_client;
					clients[i].ip = xstrdup((char *)inet_ntoa(cl_addr.sin_addr));
					strcpy(clients[i].cwd, config->default_home);

					if (pthread_create(&clients[i].pthread, NULL, (void *)client_thread, (void *) &clients[i])) {
						debug_msg(0, "%s : pthread_created(...) failed!\n", clients[i].ip);
						shutdown(new_client, 2);
						close(new_client);
					} else {
						debug_msg(0, "%s : Assigned to slot %d.\n", clients[i].ip, i);
					}
				}

			}


		}

	}

	/* never reached */
	exit(0);
}

void client_thread(client_list *client) {
	int r;

	DBG("client_thread(): client %d, thread id %ld\n", client->id, pthread_self());

	if (client_send(client->id, TELNET_WILL_ECHO) == -1) goto out;
	if (client_send(client->id, TELNET_WILL_SUPRESS_GO_AHEAD) == -1) goto out;
	if (client_send(client->id, client->cwd) == -1) goto out;
	if (client_send(client->id, " >") == -1) goto out;
	if (client_send(client->id, OUTPUT_DONE) == -1) goto out;

	for( ;; ) {
		r = client_request(client->id);

		if (r == -1) {
			debug_msg(0, "%s (%d) : disconnected.\n", client->ip, client->id);
			break;
		}

		if (r == -2) {
			debug_msg(0, "%s (%d) : client says shutdown!\n", client->ip, client->id);
			mplayerd_quit();
		}
	}


out:
	kill_client(client->id);
	pthread_exit(NULL);
}


void mplayerd_quit() {
	int err;
	int i;

	debug_msg(0, "mplayerd shutting down...\n");
	DBG("mplayerd_quit, in thread %ld\n", pthread_self());

	err = pthread_cancel(mplayer_message_pthread);

	for(i=0; i<MAX_CLIENTS; i++) {
		if (clients[i].socket <= 0) continue;
		err = client_send(i, "\r\nserver shutdown\r\n" );
		DBG("closing clients[%d], result %d\n", i, err);
		kill_client(i);
	}


//	free_instances();

	if (config->allow_ips) ll_free(config->allow_ips);
	xfree(config);

#ifdef DEBUG_MEMORY
	ll_mem_print_statistics();
#endif
	err = pthread_cancel(main_pthread);

	pthread_join(mplayer_message_pthread, NULL);
	pthread_join(main_pthread, NULL);

	DBG("Done, exiting...\n");
	exit(0);
}

char *mplayerd_version() {      
        char *buf = (void*) xmalloc(32);
        sprintf(buf, "mplayerd %s", MPLAYERD_VERSION);
        return buf;                             
}                                        
                                        
void mplayerd_help() {
	char *p;
	p = mplayerd_version();

        printf( "%s \n"                 
                "daemon to control mplayer via TCP/IP\n\n"
                "-h --help             Print this help message\n"
                "-fg --foreground      Don't fork\n"
                "-p --port port        Specify port, default %d\n"
                "-l --logfile file     Log to file default off\n"
                "-v --version          display version\n"
		"-d --debug            debug, can specify multiple times\n"
		"-c --config           specify config file\n"
		"\n"
		"You can print the entire help for commands by using flags '-d -h'\n"
		, p, DEFAULT_PORT);

	if (debug) print_command_help();

	xfree(p);
}


int parse_argv(int argc, char **argv) {
	int i;
	char *p;

	for(i=0; i<argc; i++) {

		if (!strcmp(argv[i], "-h") || !strcmp(argv[i], "--help")) {
			mplayerd_help();
			return -1;

		} else if (!strcmp(argv[i], "-fg") || !strcmp(argv[i], "--foreground")) {
			doFork = 0;

		} else if (!strcmp(argv[i], "-p") || !strcmp(argv[i], "--port")) {
			if (argv[i + 1]) {
				config->port = strtol(argv[i + 1], NULL, 0);
			} else {
				printf("port parameter required\n");
				return -1;
			}

		} else if (!strcmp(argv[i], "-l") || !strcmp(argv[i], "--logfile")) {
			logging = 1;
			if (argv[i + 1]) {
				logfile = xstrdup(argv[i + 1]);
			} else {
				printf("logfile parameter required\n");
				return -1;
			}

		} else if (!strcmp(argv[i], "-v") || !strcmp(argv[i], "--version")) {
			p = mplayerd_version();
			printf("%s\n", p);
			xfree(p);
			return -1;

		} else if (!strcmp(argv[i], "-d") || !strcmp(argv[i], "--debug")) {
			debug++;

		} else if (!strcmp(argv[i], "-c") || !strcmp(argv[i], "--config")) {
			if (argv[i + 1]) {
				strcpy(config_file, argv[i + 1]);
			} else {
				printf("config file parameter required\n");
				return -1;
			}
		}
	}
	return 0;
}

int server_setup() {
	int sock;
	struct sockaddr_in my_addr;
	int my_addrlen, yes = 1;


	memset(&my_addr, 0, sizeof(my_addr));

	my_addr.sin_family = AF_INET;
/*	my_addr.sin_addr.s_addr = INADDR_ANY; */
	my_addr.sin_addr.s_addr = inet_addr(BIND_ADDR);
	my_addr.sin_port = htons(config->port);

	my_addrlen = sizeof(my_addr);

	if ((sock = socket(AF_INET, SOCK_STREAM, 0)) == -1) {
		perror("socket");
		return -1;
	}

	if (setsockopt(sock, SOL_SOCKET, SO_REUSEADDR, &yes, sizeof(yes)) < 0) {
		perror ("setsockopt: SO_REUSEADDR: ");
		return -1;
	}


	if (bind(sock, (struct sockaddr*) &my_addr, my_addrlen) == -1) { /* Must have this on server */
		perror("bind");
		return -1;
	}

	if (listen(sock, 10) == -1) {
		perror("listen");
		return -1;
	}
	return sock;
}

int get_config_file_name(int argc, char **argv, char *file) {
	int i;  
	memset(file, 0, 1024);
	for(i=0; i<argc; i++) { 
		if (strncmp("-c", argv[i], 2) == 0 || strncmp("--config", argv[i], 8) == 0) {
			if (argv[i + 1]) strcpy(file, argv[i + 1]);
			return 0;
		}
	}
	strcpy(file, DEFAULT_CONFIG_FILE);
	return 0;
}

void print_command_help() {
	int i2 = 0, i = 0, j = 0;
	cmd_list *cmdl;
	char *p;

	printf("\n\nmplayerd commands\n"
	           "-----------------\n");

	while (1) {
		cmdl = (cmd_list *) &command_list[i2++];
		if (cmdl->cmd == NULL) break;

		printf("Syntax: %s", cmdl->cmd);

		p = cmdl->args;
		for(i=0; *(p + i); i++) {

			if (*(p + i) == '*') {
				printf(" [");
				j = 1;
				continue;
			}

			if (*(p + i) == 's') {
				printf(" $string");

			} else if (*(p + i) == 'f') {
				printf(" $file");

			} else if (*(p + i) == 'd') {
				printf(" $directory");

			} else if (*(p + i) == 'i') {
				printf(" $integer");

			} else if (*(p + i) == 'n') {
				printf(" $instance");

			} else {
				printf(" $X");
			}

			if (j) {
				printf(" ]");
				j = 0;
			}
		}
		printf("\nDescription: %s\n", cmdl->description);
		printf("%s\n\n", cmdl->help);
	}
}
